WIP: Vulkan context for VkDisplayKHR (aka: "drm" context for Vulkan)

philipl opened the pull request:
This is the Vulkan equivalent of the drm context for OpenGL, with
the big difference that it's implemented purely in terms of Vulkan
calls and doesn't actually require drm or kms.

The basic idea is to identify a display, mode, and plane on a device,
and then create a display backed surface for the swapchain. In theory,
past that point, everything is the same, and this is in fact the case
on Intel hardware. I can get a video playing on a vt.

On nvidia, naturally, things don't work that way. Instead, nvidia only
implemented the extension for scenarios where a VR application is
stealing a display from a running window system, and not for
standalone scenarios. With additional code, I've got this scenario to
work but that's a separate incremental change.

Other people have tested on AMD, and report roughly the same behaviour
as on Intel.

Note, that in this change, the VT will not be correctly restored after
qutting. The only way to restore the VT is to introduce some drm
specific code which I will illustrate in a separate change.

paulguy commented:
Works for me on a Radeon RX 580, kernel 5.3.7-arch1-1-ARCH and mesa 5.2.2.  In my case the VT isn't blank after mpv quits but just displays the last frame until I switch back to it.

https://gist.github.com/paulguy/4d04a8affe5287ae98c3bef22694393f

philipl commented:
I've come up with a way to restore the vt after qutting. Please try on amd.

paulguy commented:
Updated system to kernel 5.4.3-arch1-1 and mesa 19.3.0 and updated/rebuilt your branch and the VT does return to normal after quitting.  VT switching does something but it just pauses playback and freezes the video frame on the display but otherwise mpv's VT can be switched back to and playback continues and the VTs continue functioning normally when playback finishes.  Switching to the VT X is on is all kinds of broken and breaks keyboard input to the VTs and causes X to not come back in any meaningful way.and requires rebooting with sysrq.  Can't even reboot with CTRL+ALT+DEL.

philipl commented:
I've updated this PR to be a bit more focused.

- I've removed the vt/kms change from the series. let's consider whether the basic vulkan-only part is worth merging.
- I've added a config option to choose the display/mode/plane
- I've used a new libplacebo helper to tie the Vulkan device selection to the logic used for the actual context creation to avoid broken scenarios where display selection happens on a different device.

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,461 @@
+            continue;
+        }
+
+        planes = talloc_array(NULL, VkDisplayPlanePropertiesKHR, num_planes);
```
This function leaks a lot of memory, especially in the failure cases. It's also confusing to track the lifetime of all those allocations. I suggest grouping all these allocations under a single `tmp` object that you then free at the end the function.

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,461 @@
+    talloc_free(modes);
+    talloc_free(planes);
+    talloc_free(planes_to_displays);
+    return ret;
```
You should also clean up the vulkan instance

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,461 @@
+    }
+
+    res = vkGetPhysicalDeviceDisplayPropertiesKHR(device, &num, NULL);
+    if (res != VK_SUCCESS) {
```
FWIW these cases are basically impossible, might as well not even bother testing this return value. Just make sure to initialize `num` to 0 and nothing will go wrong even if it hypothetically fails somehow.

haasn commented:
I should have been asleep 6 hours ago. I'll look it more some other time

philipl commented:
> I should have been asleep 6 hours ago. I'll look it more some other time

Thanks @haasn. I've updated to reflect your feedback.

And for anyone who is interested, I've got the kms changes (and will add the nvidia changes) on [another branch](https://github.com/philipl/mpv/tree/vulkan_display_wip).

philipl commented:
We're coming up on a year since I last touched any of this. Do we think this is worth merging? The console restoration via kms is still WIP but I understand what needs to be done to complete that work, and I still want to merge the pure vulkan code first.

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,423 @@
+    char *display_spec;
+};
+
+static bool parse_display_spec(struct mp_log *log,
```
This function seems weirdly written for a number of reasons. It seems to contain logic for dealing with partial strings, but a string like "0" would just get rejected by the function unless it contains a `:` as well.

I recommend finding another way to write this in a way that's a less confusing mess of conditionals, and does the right thing on partial strings. Whatever that "right thing is", I'm not sure what the intended logic is, but on my system at least, specifying the mode has the highest priority, followed by the display/plane, because there's only one of the latter. So I'd expect this being in the order `[[display:]plane:]mode`, with just e.g. `5` being interpreted as `0:0:5`. :woman_shrugging:

Is this a common form of identifier or something you made up?

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,423 @@
+        if (plane_ptr) {
+            parsed_mode = atoi(mode_ptr + 1);
+        }
+        if (*(plane_ptr + 1) != '\0') {
```
Segfault if `plane_ptr` is NULL. As written, this needs to be inside the previous `if`. (Also, I think `plane_ptr[1]` is more readable than `*(plane_ptr + 1)`)

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,423 @@
+
+        ret = true;
+    }
+    mp_dbg(log, "Parsed Display Spec: %d, %d, %d\n",
```
Nit: Should probably output this in the same format as it's parsed in, i.e. `%d:%d:%d`.

haasn commented on video/out/vulkan/context_display.c:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    char *spec = bstrto0(NULL, param);
+    if (!parse_display_spec(log, spec, NULL, NULL, NULL)) {
```
Should the enumeration code from `print_display_info` be merged into this function so that you can also verify that the parsed integers correspond to really existing modes?

haasn reviewed (changes requested):
Seems reasonable, with some minor issues/notes. Also, missing documentation.

haasn commented:
I can't help but wonder if the whole DRM restoring thing shouldn't be done in the vulkan driver. Have you tried raising this issue with mesa? It seems like we're being asked to do things we either shouldn't, or can't.

philipl commented:
Thanks. I need to find time to get back to this, but I will rework the command line to be three separate arguments, which is how the drm context works, so that we don't have to deal with any weird parsing.

philipl commented:
I've updated the change to take the display/mode/plane parameters as separate arguments. Validation remains an unsolved problem. You can't validate a single option without knowing what the other values are, so that really means you have to validate at actual init time. I don't see a reasonable way around this.

Still need to add documentation.

philipl commented:
Added documentation of options.

philipl commented:
Updated to refactor enumeration of the various VkDisplay entities to share between init and help.

philipl commented:
I've updated the formatting and indentation. @haasn Are we good to go?
